base/sparse/sparsematrix.jl

```diff
@@ -973,17 +973,89 @@ function getindex_I_sorted{Tv,Ti}(A::SparseMatrixCSC{Tv,Ti}, I::AbstractVector,
     # Sorted vectors for indexing rows.
     # Similar to getindex_general but without the transpose trick.
     (m, n) = size(A)
+    minj, maxj = extrema(J)
+    ((I[1] < 1) || (I[end] > m) || (minj < 1) || (maxj > n)) && BoundsError()
+
     nI = length(I)
-    nJ = length(J)
+    avgM = div(nnz(A),m)
+    # heuristics based on experiments
+    alg = ((nI - avgM) > 2^8) ? 1 :
+          ((avgM - nI) > 2^10) ? 0 : 2
+
+    (alg == 0) ? getindex_I_sorted_binary_A(A, I, J) :
+    (alg == 1) ? getindex_I_sorted_binary_I(A, I, J) :
+    getindex_I_sorted_linear(A, I, J)
+end
+
+function getindex_I_sorted_binary_A{Tv,Ti}(A::SparseMatrixCSC{Tv,Ti}, I::AbstractVector, J::AbstractVector)
+    const nI = length(I)
+    const nJ = length(J)
 
     colptrA = A.colptr; rowvalA = A.rowval; nzvalA = A.nzval
     colptrS = Array(Ti, nJ+1)
     colptrS[1] = 1
-    nnzS = 0
 
-    # Form the structure of the result and compute space
-    for j = 1:nJ
-        @inbounds col = J[j]
+    offI = I[1]-1
+
+    ptrS = 1
+    # determine result size
+    @inbounds for j = 1:nJ
+        col = J[j]
+        ptrI::Int = 1 # runs through I
+        ptrA::Int = colptrA[col]
+        stopA::Int = colptrA[col+1]-1
+        while ptrI <= nI
+            rowI = I[ptrI]
+            ptrI += 1
+            (rowvalA[ptrA] > rowI) && continue
+            ptrA = searchsortedfirst(rowvalA, rowI, ptrA, stopA, Base.Order.Forward)
+            (ptrA <= stopA) || break
+            if rowvalA[ptrA] == rowI
+                ptrS += 1
+            end
+        end
+        colptrS[j+1] = ptrS
+    end
+
+    rowvalS = Array(Ti, ptrS-1)
+    nzvalS  = Array(Tv, ptrS-1)
+
+    # fill the values
+    ptrS = 1
+    @inbounds for j = 1:nJ
+        col = J[j]
+        ptrI::Int = 1 # runs through I
+        ptrA::Int = colptrA[col]
+        stopA::Int = colptrA[col+1]-1
+        while ptrI <= nI
+            rowI = I[ptrI]
+            if rowvalA[ptrA] <= rowI
+                ptrA = searchsortedfirst(rowvalA, rowI, ptrA, stopA, Base.Order.Forward)
```

Comment:
It does make it slightly faster(~twice). But since we got quite a speedup from other mechanisms, and marking `searchsorted` inlined may affect other places, I didn't include it in the PR.